File: app/src/main/java/com/dmitrybrant/android/mandelbrot/simple/ColorScheme.kt
package com.dmitrybrant.android.mandelbrot.simple

import android.graphics.Color

object ColorScheme {
    private var colorSchemes = mutableListOf<IntArray>()

    fun initColorSchemes() {
        colorSchemes = ArrayList()
        colorSchemes.add(createColorScheme(intArrayOf(Color.BLUE, Color.GREEN, Color.RED, Color.BLUE), 256))
        colorSchemes.add(createColorScheme(intArrayOf(Color.YELLOW, Color.MAGENTA, Color.BLUE, Color.GREEN, Color.YELLOW), 256))
        colorSchemes.add(createColorScheme(intArrayOf(Color.WHITE, Color.BLACK, Color.WHITE), 256))
        colorSchemes.add(createColorScheme(intArrayOf(Color.BLACK, Color.WHITE, Color.BLACK), 256))
        colorSchemes.add(intArrayOf(Color.BLACK, Color.WHITE))
    }

    fun getColorSchemes(): List<IntArray> {
        return colorSchemes
    }

    fun getShiftedScheme(colors: IntArray, shiftAmount: Int): IntArray {
        val shifted = IntArray(colors.size)
        for (i in colors.indices) {
            shifted[i] = colors[(i + shiftAmount) % colors.size]
        }
        return shifted
    }

    private fun createColorScheme(colorArray: IntArray, numElements: Int): IntArray {
        val elementsPerStep = numElements / (colorArray.size - 1)
        val colors = IntArray(numElements)
        var r = 0f
        var g = 0f
        var b = 0f
        var rInc = 0f
        var gInc = 0f
        var bInc = 0f
        var cIndex = 0
        var cCounter = 0
        for (i in 0 until numElements) {
            if (cCounter == 0) {
                b = (colorArray[cIndex] and 0xff.toFloat().toInt()).toFloat()
                g = (colorArray[cIndex] and 0xff00 shr 8.toFloat().toInt()).toFloat()
                r = (colorArray[cIndex] and 0xff0000 shr 16.toFloat().toInt()).toFloat()
                if (cIndex < colorArray.size - 1) {
                    bInc = ((colorArray[cIndex + 1] and 0xff).toFloat() - b) / elementsPerStep.toFloat()
                    gInc = ((colorArray[cIndex + 1] and 0xff00 shr 8).toFloat() - g) / elementsPerStep.toFloat()
                    rInc = ((colorArray[cIndex + 1] and 0xff0000 shr 16).toFloat() - r) / elementsPerStep.toFloat()
                }
                cIndex++
                cCounter = elementsPerStep
            }
            colors[i] = -0x1000000 or (r.toInt() shl 16) or (g.toInt() shl 8) or b.toInt()
            b += bInc
            g += gInc
            r += rInc
            if (b < 0f) {
                b = 0f
            }
            if (g < 0f) {
                g = 0f
            }
            if (r < 0f) {
                r = 0f
            }
            if (b > 255f) {
                b = 255f
            }
            if (g > 255f) {
                g = 255f
            }
            if (r > 255f) {
                r = 255f
            }
            cCounter--
        }
        return colors
    }
}

File: ios/MandelbrotApp/Shared/ColorScheme.swift
import Foundation

struct ColorSchemeManager {
    // Android ARGB color constants
    private static let BLUE:   UInt32 = 0xFF0000FF
    private static let GREEN:  UInt32 = 0xFF00FF00
    private static let RED:    UInt32 = 0xFFFF0000
    private static let YELLOW: UInt32 = 0xFFFFFF00
    private static let MAGENTA: UInt32 = 0xFFFF00FF
    private static let WHITE:  UInt32 = 0xFFFFFFFF
    private static let BLACK:  UInt32 = 0xFF000000

    private(set) static var colorSchemes: [[UInt32]] = []

    static func initColorSchemes() {
        colorSchemes = [
            createColorScheme(colors: [BLUE, GREEN, RED, BLUE], numElements: 256),
            createColorScheme(colors: [YELLOW, MAGENTA, BLUE, GREEN, YELLOW], numElements: 256),
            createColorScheme(colors: [WHITE, BLACK, WHITE], numElements: 256),
            createColorScheme(colors: [BLACK, WHITE, BLACK], numElements: 256),
            [BLACK, WHITE]
        ]
    }

    static func getShiftedScheme(colors: [UInt32], shiftAmount: Int) -> [UInt32] {
        let count = colors.count
        var shifted = [UInt32](repeating: 0, count: count)
        for i in 0..<count {
            shifted[i] = colors[(i + shiftAmount) % count]
        }
        return shifted
    }

    private static func createColorScheme(colors colorArray: [UInt32], numElements: Int) -> [UInt32] {
        let elementsPerStep = numElements / (colorArray.count - 1)
        var colors = [UInt32](repeating: 0, count: numElements)
        var r: Float = 0
        var g: Float = 0
        var b: Float = 0
        var rInc: Float = 0
        var gInc: Float = 0
        var bInc: Float = 0
        var cIndex = 0
        var cCounter = 0

        for i in 0..<numElements {
            if cCounter == 0 {
                b = Float(colorArray[cIndex] & 0xFF)
                g = Float((colorArray[cIndex] & 0xFF00) >> 8)
                r = Float((colorArray[cIndex] & 0xFF0000) >> 16)
                if cIndex < colorArray.count - 1 {
                    bInc = (Float(colorArray[cIndex + 1] & 0xFF) - b) / Float(elementsPerStep)
                    gInc = (Float((colorArray[cIndex + 1] & 0xFF00) >> 8) - g) / Float(elementsPerStep)
                    rInc = (Float((colorArray[cIndex + 1] & 0xFF0000) >> 16) - r) / Float(elementsPerStep)
                }
                cIndex += 1
                cCounter = elementsPerStep
            }
            colors[i] = 0xFF000000 | (UInt32(r) << 16) | (UInt32(g) << 8) | UInt32(b)
            b += bInc
            g += gInc
            r += rInc
            b = min(max(b, 0), 255)
            g = min(max(g, 0), 255)
            r = min(max(r, 0), 255)
            cCounter -= 1
        }
        return colors
    }
}
